docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,263 @@
+# using an extra IP via dhcp sounds like the best solution.
+# We can leave all other routes which belong to $DEV untouched.
+# No matter what we do, we don't want any routes modified.
+ip addr add $DEV_SUBNET.13/24 dev mv4 noprefixroute
+
+# normally we already have the netork namespace of libvirt given.
+ip netns add ns1
+
+# create the macvlan link attaching it to the parent host $DEV
+ip link add mv1 link $DEV type macvlan mode bridge
+
+# move the new interface mv1 to the new namespace
+ip link set mv1 netns ns1
+
+# bring the interface up
+# ns1 will later be used by libvirt to attach macvtaps to
+ip netns exec ns1 ip link set dev mv1 up
+
+
+```
+
+After that invocation, we have a host macvlan, with an IP address assigned. The
+CNI dhcp plugin comes with a minimalistic dhcp client, which we can use to do
+the DHCP renewals. An important side note here is, that the minimalistic daemon
+returns the acquired IP alog side with routing rules suggested by the DHCP
+server, but they are not applied, so the host is not modified.
+
+#### Prepare the VM device
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: HostNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+For every `HostNetwork`, `virt-handler` let's libvirt create a new macvtap
+attached to the one prepared macvlan in the initial setup step in the libvirt
+namespace.  Ee will use the CNI `dhcp` IPAM plugin with a provided or generated
+MAC address. Based on this MAC address and the IPAM plugin invocation we end up
+with the correct IP and MAC pair. The MAC address will then be used by libvirt
+as the mactap address. The IP can be used to add a route, to allow routing
+between the new guest and the host.
+
+Taking the macvlan name and the MAC address, we can now transform the
+HostNetwork entry into a libvirt XML fragment:
+
+```xml
+<interface type='direct'>
+  <mac address="fe:54:00:f7:78:6d"/>
+  <source dev='macvlan1' mode='bridge'/>
+</interface>
+```
+
+When the guest now tries to acquire an IP address with this mac address, it
+will get the same IP address.
+
+```bash
+# Add an extra route which allows accessing the $DEV_IP through the master macvlan.
+# We need one of these routes per device.
+ip route add $DEV_SUBNET.11/32 via $DEV_SUBNET.13 dev mv4
+
+# ping from container to host
+# done by the ipam dhcp client
+ip netns exec ns1 ping $DEV_SUBNET.13
+```
+
+In this scenario, we can keep the IP of our main interface in place, and only
+have to add one additional route per acquired IP for a VM. All existing routes
+can stay untouched. In theory, moving the IP of the main interface to `mv4`
+would also be possible, but that would require extensive route rewrites which
+will likely cause conflicts with other components involved.
+
+### Bridge Scenario (tap)
+
+Normally with the Kubernetes overlay networks, bridges are not required or
```

Comment:
> I am not familiar with any k8s network solution that does not use bridges, I think you meant that the bridges used are never attached to the host NIC directly.

yes, that is what I meant.

> I would argue that we do need to own the host network, at least at the stage when we prepare the network.
> It may be some other service that owns the host network (like a network manager) and we should be fine with assuming that the bridge existence is a pre requisite for kubevirt.

If we own it, we can tell people or force people to use bridges. However, with the macvlan/macvtap approach we can make this feature work on almost all installations, without messing up peoples host network. We only add a specific route on VM start, and we remove a specific route on VM stop. If a nested VM  does not need to talk to the machine it is scheduled on, not even that is necessary.